pixel_clusterizer/cluster_functions.py

```diff
@@ -1,182 +1,297 @@
+        charge_correction = 1
+    else:
+        charge_correction = 0
 
     # Temporary variables that are reset for each cluster or event
-    actual_event_number, actual_event_hit_index, actual_event_cluster_index, actual_cluster_id, max_cluster_charge, next_cluster_id, actual_cluster_size, actual_cluster_hit_index, seed_index = 0, 0, 0, 0, 0, 0, 0, 0, 0
-    actual_cluster_hit_indices = np.zeros(shape=max_n_cluster_hits, dtype=np.int16) - 1  # The hit indices of the actual cluster, -1 means not assigned
+    start_event_hit_index = 0
+    start_event_cluster_index = 0
+    cluster_size = 0
+    event_number = hits[0]['event_number']
+    event_cluster_index = 0
 
     # Outer loop over all hits in the array (referred to as actual hit)
-    for i in range(hits.shape[0]):
-        if i >= n_hits:
-            break
+    for i in range(total_hits):
 
         # Check for new event and reset event variables
-        if _new_event(hits[i]['event_number'], actual_event_number):
-            _finish_event(hits, cluster, is_seed, n_cluster, cluster_size, cluster_id, actual_event_hit_index, i + 1, next_cluster_id, actual_event_cluster_index)
-            actual_event_hit_index = i
-            actual_event_cluster_index = actual_event_cluster_index + next_cluster_id
-            actual_event_number = hits[i]['event_number']
-            next_cluster_id = 0  # First cluster has ID 1
-
-        # Reset temp array with hit indices of actual cluster for the next cluster
-        _reset_cluster_hit_indices(actual_cluster_hit_indices, actual_cluster_size)
-        actual_cluster_hit_index = 0
-
-        # Check if actual hit is already asigned to a cluster, if not define new actual cluster containing the actual hit as the first hit
-        if cluster_id[i] != -1:  # Hit was already assigned to a cluster in the inner loop, thus skip actual hit
+        if _new_event(hits[i]['event_number'], event_number):
+            _finish_event(
+                hits=hits,
+                cluster=cluster,
+                start_event_hit_index=start_event_hit_index,
+                stop_event_hit_index=i,
+                start_event_cluster_index=start_event_cluster_index,
+                stop_event_cluster_index=start_event_cluster_index + event_cluster_index)
+
+            start_event_hit_index = i
+            start_event_cluster_index = start_event_cluster_index + event_cluster_index
+            event_number = hits[i]['event_number']
+            event_cluster_index = 0
+
+        if assigned_hit_array[i] > 0:  # Hit was already assigned to a cluster in the inner loop, thus skip actual hit
             continue
 
-        # Omit hits with charge > max_cluster_hit_charge
-        if hits[i]['charge'] > max_cluster_hit_charge:
+        if not _hit_ok(
+                hit=hits[i],
+                min_hit_charge=min_hit_charge,
+                max_hit_charge=max_hit_charge) or (disabled_pixels.shape[0] != 0 and _pixel_masked(hits[i], disabled_pixels)):
```

Comment:
Because it is a numba bug/feature that evaluate
```
(disabled_pixels.shape[0] != 0 and _pixel_masked(hits[i], disabled_pixels)
``` 
fully when this line is buried in another function. This makes the clustering slow.